Add lint, build and test jobs to CI

tharvik opened the pull request:
Mine first PR to disco, tell me if I missed anything :grin: 

Takes care of #225

It triggers upon merging from `develop` -> `production` and on PR creation.
 
> * Run server tests
> * Build server docker
> * Run client tests

Done and
* add some caching while at it
* change the client's `npm run test` to also start the server


> * Run server docker

Hum, I'm using the one build from sources for the test, is it better to start the docker one?

> * Build and update to github pages

The deploy workflow already takes care of that. I changed the trigger to do it iff the `lint-build-test` passes.

> * Update google app engine (server)

I don't know what are the credentials for GAE.
One can use [Google's GitHub Action](https://github.com/google-github-actions/deploy-appengine)  for that, feel free to push to this PR if you have theses.

> Allow only a PR to be pushed for review if it passes auto tests (PR-draft would still work without tests)

Enforcing this need an admin change to the setting of the repo. It would be best to enforce it after merging this PR.

tharvik committed:
github: add lint, test and build workflow

Closes: #225

tharvik committed:
server: fix lint

tharvik committed:
client: start server on `npm run test`

tharvik committed:
github/deploy: trigger when lint-test-build pass

Nacho114 commented:
We feel honored to receive your first PR! :)

`Hum, I'm using the one build from sources for the test, is it better to start the docker one?`

Good question, maybe it is best to run the docker image we build? We build it during the process as you point out, so we can verify it builds correctly, might as well go the extra step and check it works.

Nacho114 commented on mobile-browser-based-version/package.json:
```diff
@@ -7,7 +7,7 @@
     "build": "vue-cli-service build",
     "lint": "vue-cli-service lint",
-    "test": "env TS_NODE_COMPILER_OPTIONS='{\"module\": \"commonjs\" }' mocha --exit -r ts-node/register --file tests/setup.ts 'tests/**/*.ts'",
+    "test": "( cd server && npm start ) & env TS_NODE_COMPILER_OPTIONS='{\"module\": \"commonjs\" }' mocha --exit -r ts-node/register --file tests/setup.ts 'tests/**/*.ts' && kill $! && wait",
```
the server takes a bit of time to build (2-3s), maybe in the future it might be longer, could that be a problem? e.g. the tests start before the server finishes building.

tharvik replied:
Right, I'm now using a script to ensure that it is started before running the tests: `with_server`

Nacho114 commented:
`I don't know what are the credentials for GAE.`

I have them, I can take care of this.

Nacho114 commented:
Ah I just saw, there is a GitHub section now with:

`All checks have passed`

Very cool stuff <3

tharvik commented:
>> I'm using the one build from sources for the test, is it better to start the docker one?
>
> Good question, maybe it is best to run the docker image we build? We build it during the process as you point out, so we can verify it builds correctly, might as well go the extra step and check it works.

I added a script, `with_server`, that ensure that a server is answering before running the given command, then takes care of stopping it. It is used by `npm run test`.

>> I don't know what are the credentials for GAE.
>
> I have them, I can take care of this.

Alright, fell free to push to this PR :)

Commit:
added deploy-server.ymp script

Commit:
renamed names to follow dash-case convention

Nacho114 commented:
I added the gcloud deploy automation script, added the corresponding gcloud credentials to github secrets + renamed the names in the yml files to follow dash-case. Let me know if it looks good / naming is better like this.

tharvik commented on .github/workflows/deploy-server.yml:
```diff
@@ -0,0 +1,33 @@
+    runs-on: ubuntu-latest
+    defaults:
+      run:
+        working-directory: ${{ env.server_dir }}
```
This is only set for `run` steps, if you want to use it for `google-github-actions/deploy-appengine`, you can use its [`working_directory` input](https://github.com/marketplace/actions/deploy-to-app-engine#inputs)

Nacho114 commented on mobile-browser-based-version/package.json:
```diff
@@ -7,7 +7,8 @@
     "build": "vue-cli-service build",
     "lint": "vue-cli-service lint",
-    "test": "env TS_NODE_COMPILER_OPTIONS='{\"module\": \"commonjs\" }' mocha --exit -r ts-node/register --file tests/setup.ts 'tests/**/*.ts'",
+    "blam": "false; true",
```
what is blam for?

tharvik replied:
hoo, good catch, it was a debug line; removing it

tharvik commented on .github/workflows/deploy-server.yml:
```diff
@@ -0,0 +1,33 @@
+        uses: actions/checkout@v2
+
+      - name: deploy
+        uses: google-github-actions/deploy-appengine@v0.2.0
```
Is there a reason behind using an old version?

```suggestion
        uses: google-github-actions/deploy-appengine@v0.8.0
```

btw, to ensure that it works, I usually force it to run, by branching of and changing the trigger to `on: push`

Nacho114 replied:
thanks for upgrading it, the reason for old version: copy pasted code. How to test: thanks that's a good tip, will test it out.

tharvik committed:
client/npm: rm debug script

tharvik commented:
> I added the gcloud deploy automation script, added the corresponding gcloud credentials to github secrets + renamed the names in the yml files to follow dash-case. Let me know if it looks good / naming is better like this.

Indeed, that more consistent.

Commit:
fixed working_directory

Commit:
working (has been tested) deploy-server.yml
